### PR TITLE
EES-4932 Bulk import: allow users to import multiple data sets from a single zip file

### PR DESCRIPTION
**Pull request by mmyoungman**
This PR adds the ability to upload multiple data sets at once when importing new data sets.

image

Bulk zip files contain a `dataset_names.csv` which includes the data set title for each data set, and a file base name. The base name is used to find the actual data files also included in the zip, which it expects to be named `[basename].csv` and `[basename].meta.csv`

Examples of valid and invalid bulk zip files can be found in `src/GovUk....Admin.Tests/Resources`.

The new `ReleaseController#UploadAsBulkZip` method can return multiple errors rather than one error at a time as we've done in the past. This is because bulk zips are liable to be large files and it's easy to make a mistake when creating them, so this work was done on the backend as a convenience for the user who might otherwise have to upload-fix-upload-fix the same file multiple times.

Even though this work to support multiple errors has been done in the backend, the frontend still returns only one error at a time - mostly just so we can get this feature out quicker. A ticket will be created to add support for multiple errors to the frontend.

This work also includes a refactor of existing methods in `ReleaseDataFileService`'s `Upload` and `UploadAsZip`, mostly how they're validated. `UploadAsZip` previously wasn't validating all the same cases as `Upload`.

All upload methods now return multiple errors - which was necessary to create reusable validation methods for all types of upload - `FileUploadsValidatorService#ValidateDataSetFilesForUpload` is called by them all and handles most validation for a specific data set.

`DataImportService#Import` now supports all types of upload. The Data Processor basically treats DataZips and BulkDataZips as the same, treating each "import" as an individual data set. BulkDataZips call the Data Processor multiple times, once per data set.

### Other changes
- `ReleaseDataFileService#Delete` has been updated to not remove the source zip file if other data sets still exist with that zip as a source.
- `FileTypeService` has been changed a lot. This was started as I wanted `IsValidCsv` to accept a stream rather than a stream provider, and then basically cascaded into a major refactoring.
- Removed an unused `ErrorViewModel` file from Admin - we're using the one in Common.
- I've done a lot of renaming - mostly subject/dataSetName to dataSetTitle.

### CODE REVIEW

**Comment by mmyoungman** [comment]
> I think it's probably too late to change now but it seems like we have a lot of extra complexity for handling zip files and bulk zip files [...]

I did mostly use the same pattern that existing DataZips used for BulkZips. If we didn't do that, then it would have involed refactoring the existing DataZip functionality before starting the BulkZip work.

I agree with that if we instead uploaded files to the Data Processor and that handled all validation and importing that would be simpler, but in terms of getting from where we were initially to having bulk imports, that extra refactoring adds a lot of work - and I mean *a lot* of work, in my opinion - and I don't think it's justified.

If there is one way I'm inclined to think you may have a point is that maybe refactoring it now would make implementing some future requirement much easier in a way that is difficult to foresee. The reason I discount that is because that's a what-if, while the time saved in implementation is a concrete saving we gain right now. Although not saying you couldn't be right - you have much more experience than me and I'd be a fool to totally discount your opinion :)

I don't think the bulk zip work puts a refactor like you want totally out of the question - but I agree it puts it further out of reach. That said, how much extra work would it be to do later compared to now? While I think refactoring now would have been more efficient than refactoring later, but I'm not sure it's a huge saving.

> The validation which is spread across the Admin and Data Processor is also pretty complicated.

If this ends up with Admin dealing with all the validation - say with the screener added - and the Processor with importing, I'm inclined to think that is okay. It doesn't feel to me like the complexity is will become out of hand - of the stuff I touched, I found it fairly easy to follow. 

If we had extra time to spend, there are many other things where I think the complexity is getting out of control - something like untangling the publication of methodologies and releases (which in turn made redirects a nightmare) would be a much better place to spend extra time, I think.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs [hunk 1]
```diff
@@ -137,16 +137,29 @@ public async Task<ActionResult<List<DataFileInfo>>> GetDataFileInfo(Guid release
...
-                    subjectName: subjectName)
+                    dataSetTitle: dataSetTitle,
+                    replacingFileId: replacingFileId)
+                .HandleFailuresOrOk();
+        }
+
+        [HttpPost("release/{releaseVersionId:guid}/bulk-zip-data")]
+        [Produces("application/json")]
+        [ProducesResponseType(200)]
+        [ProducesResponseType(404)]
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:156]: Removed all `Produces` and `ProducesResponseType` attributes out of controller actions in `ReleasesController.cs`.

I was tempted to do a codebase-wide cleanse, but probably better not pollute this PR, given how large it is.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs [hunk 2]
```diff
@@ -137,16 +137,29 @@ public async Task<ActionResult<List<DataFileInfo>>> GetDataFileInfo(Guid release
...
+        [RequestSizeLimit(int.MaxValue)]
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:157]: > Is the intention to limit the request size to 2Gb here?

Frankly, I was just copying the other endpoints. I could switch to `DisableRequestSizeLimit` as you suggest, but it the current attribute these has been in place for a long time and I'm wary of changing something that could potentially break something, without spending the appropriate time investigating it.

My inclination is to create a ticket to do this separately. But if you're confident in the change, I'm happy to do it.

> Also just wondering if you have done any testing of large zip file uploads for this new bulk upload feature?

I haven't - I'm inclined to leave that to the testers. If this goes to the dev branch and we do have a problem, a small migration will be required to revert the change, but otherwise it should be fine.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Processor/Services/ValidatorService.cs [hunk 3]
```diff
@@ -90,13 +90,13 @@ public class ValidatorService : IValidatorService
...
-                    
+
             var metaFileStreamProvider = () => _privateBlobStorageService.StreamBlob(PrivateReleaseFiles,
                 import.MetaFile.Path());
 
             return await
-                ValidateCsvFileType(import.MetaFile, metaFileStreamProvider, true)
-                    .OnSuccess(() => ValidateCsvFileType(import.File, dataFileStreamProvider, false))
+                ValidateCsvFileType(metaFileStreamProvider, true)
+                    .OnSuccess(() => ValidateCsvFileType(dataFileStreamProvider, false))
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Processor/Services/ValidatorService.cs:99]: I've added `isMetaFile:` label to `ValidateCsvFileType` calls and removed the superfluous `isMetaFile` argument from `ValidateMetadataFile`.

File: src/GovUk.Education.ExploreEducationStatistics.Content.Services/DataSetFileService.cs [hunk 4]
```diff
@@ -283,8 +282,9 @@ private async Task<DataSetFileCsvPreviewViewModel> GetDataCsvPreview(ReleaseFile
             containerName: BlobContainers.PublicReleaseFiles,
             path: releaseFile.PublicPath());
 
-        using var dataFileReader = new StreamReader(await datafileStreamProvider.Invoke());
-        using var csvReader = new CsvReader(dataFileReader, CultureInfo.InvariantCulture);
+        await using var stream = await datafileStreamProvider.Invoke();
+        using var streamReader = new StreamReader(stream);
+        using var csvReader = new CsvReader(streamReader, CultureInfo.InvariantCulture);
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Services/DataSetFileService.cs:287]: This was more for clarity of knowing for a fact that all streams will be closed due to being in a using block.

I think you said in a previous PR that you had to check whether certain streams were getting closed - my thought is this makes it crystal clear.

It seems like calling close on `csvReader` and `dataFileReader` was enough, but for completeness, I'd prefer to know for a fact it gets called on the `datafileStreamProvider.Invoke()` stream too.

So tl;dr the change is for improved code glancability.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs [hunk 5]
```diff
@@ -194,27 +89,20 @@ private string GetMimeTypeUsingMimeProject(Stream stream)
...
+            return magic.Read(streamReader.BaseStream, 1024); // magic.Read handles if buffer size is < 1024
         }
 
         private bool IsMimeTypeNullOrZip(string mimeType)
         {
-            return mimeType == null ||
-                   mimeType == "application/octet-stream" ||
-                   mimeType == "application/zip";
+            return mimeType == "application/octet-stream"
+                   || mimeType == "application/zip";
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs:105]: I think I initially removed this because Rider said it was redundant. Looking into it more, `GuessMagicInfo` returns a non-nullable string, and the `Magic` lib we're calling always returns a string and throws an exception otherwise:

```
    public string Read(byte[] buffer, int bufferSize)
    {
      this.ThrowIfDisposed();
      int length = buffer.Length < bufferSize ? buffer.Length : bufferSize;
      return Marshal.PtrToStringAnsi(MagicNative.magic_buffer(this._magic, buffer, length)) ?? throw new MagicException(this.LastError);
    }
```

Since the method is only being called once, I've inlined it and added a comment for clarity.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs [hunk 6]
```diff
@@ -194,27 +89,20 @@ private string GetMimeTypeUsingMimeProject(Stream stream)
             return await stream.GetFileTypeAsync();
         }
 
-        private string GetMimeEncoding(Stream stream)
-        {
-            return GuessMagicInfo(stream, MagicOpenFlags.MAGIC_MIME_ENCODING);
-        }
-
-        private string GuessMagicInfo(Stream fileStream, MagicOpenFlags flag)
+        private string GuessMagicInfo(Stream fileStream,  MagicOpenFlags flag)
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs:92]: I've not been commenting on these kinds of issues in PRs recently. My thought is they don't cause any issues, and the next person who makes changes to code in that area will see it and fix it.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs [hunk 7]
```diff
@@ -194,27 +89,20 @@ private string GetMimeTypeUsingMimeProject(Stream stream)
...
-            using var reader = new StreamReader(fileStream);
-
             var dbPath = _configuration.GetValue<string>("MagicFilePath");
             var magic = new Magic(flag, dbPath);
 
-            var bufferSize = reader.BaseStream.Length >= 1024 ? 1024 : (int) reader.BaseStream.Length;
-            return magic.Read(reader.BaseStream, bufferSize);
+            // I have no idea why wrapping the Stream as a StreamReader is necessary, but it is.
+            using var streamReader = new StreamReader(fileStream);
+            return magic.Read(streamReader.BaseStream, 1024); // magic.Read handles if buffer size is < 1024
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs:99]: In the two places `GuessMagicInfo` is called, the stream is opened immediately preceding calling the method.

So I've actually changed the arguments of `GuessMagicInfo` to accept the `IFormFile` and then open the stream inside the function to guarantee it's position.

After that change, I tried passing the `fileStream` created in `GuessMagicInfo` to `magic.Read`, and the unit tests passed... no idea why inlining the stream creation would cause this...

I'll give it a quick manual test too.

[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs:99]: I tried moving it back to see if it would fail, it didn't, so I'm inclined to think you're suspicion of the stream not being at position 0 is correct - or it was previously and after a refactor that never happens anymore. So the `new StreamReader` can be removed.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs [hunk 8]
```diff
@@ -116,9 +116,9 @@ public async Task<ActionResult<List<DataFileInfo>>> GetDataFileInfo(Guid release
         [ProducesResponseType(404)]
         [RequestSizeLimit(int.MaxValue)]
         [RequestFormLimits(ValueLengthLimit = int.MaxValue, MultipartBodyLengthLimit = int.MaxValue)]
-        public async Task<ActionResult<DataFileInfo>> AddDataFilesAsync(Guid releaseVersionId,
+        public async Task<ActionResult<DataFileInfo>> Upload(Guid releaseVersionId,
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:119]: > I think the name Upload doesn't give much away in the context of the ReleasesController. How about UploadDataSet? Similar comment regarding the other data set file operations in this controller.

Renamed the Upload* methods in `ReleasesController`. I've not renamed the `releaseDataFileService` methods - but can if you think that's worthwhile.

> Did you consider moving data set file methods out to a new controller because it's already pretty busy with core release methods?

I may have done? :)

I'll leave them for now, but just say if you want them moved.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Model/ArchiveDataSetFile.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        DataFileSize = dataFileSize;
+        MetaFilename = metaFilename;
+        MetaFileSize = metaFileSize;
+    }
+
+    private string _title;
+    public string Title
+    {
+        get => _title;
+        set => _title = value.Trim();
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Model/ArchiveDataSetFile.cs:23]: I've added a `Trim` to  where you suggest for bulk uploads. This won't affect data zip uploads anymore, but I think that's fine.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Interfaces/IReleaseDataFileService.cs [hunk 10]
```diff
@@ -31,15 +31,21 @@ public interface IReleaseDataFileService
...
             List<Guid> fileIds);
 
-        Task<Either<ActionResult, DataFileInfo>> Upload(Guid releaseVersionId,
+        Task<Either<ActionResult, DataFileInfo>> Upload(
+            Guid releaseVersionId,
             IFormFile dataFormFile,
             IFormFile metaFormFile,
-            Guid? replacingFileId = null,
-            string? subjectName = null);
+            string? dataSetTitle = null,
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Interfaces/IReleaseDataFileService.cs:38]: `Upload` and `UploadAsZip` must be provided either a `dataSetTitle` or a `replacingFIleId`. If it's an original import, it uses the `dataSetTitle`. If it's a replacement, it uses the `dataSetTitle` of `ReleaseFile` being replaced.

`GetReleaseVersionDataSetTitle` ensures that either one or the other is provided in the arguments.

If your query is specifically "Why make them optional arguments?", I think that's just because that's what the code was originally. I think both are always supplied (in non-test code, at least) when the methods are called. 

I've assumed removing the default value was the change you had in mind so I've removed the default value for both args.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/ReleasesControllerTests.cs [hunk 11]
```diff
@@ -69,16 +68,16 @@ public async Task AddDataFilesAsync_UploadsTheFiles_Returns_Ok()
                 .Setup(service => service.Upload(_releaseVersionId,
                     dataFile,
                     metaFile,
-                    null,
-                    "Subject name"))
+                    "Subject name",
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/ReleasesControllerTests.cs:71]: > As you are changing this I think you may as well change the "Subject name" to something else.

Done

>  I'm not sure these tests are adding much value though.

Agreed - that's why I didn't bother adding any tests here for UploadDataSetAsZip / UploadDataSetsAsBulkUpload. I was actually tempted to remove them completely, but I ended up leaving them.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataArchiveValidationService.cs [hunk 12]
```diff
@@ -18,99 +23,274 @@ namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
...
-            };
-
-        public DataArchiveValidationService(IFileTypeService fileTypeService)
+        public DataArchiveValidationService(
+            IFileTypeService fileTypeService,
+            IFileUploadsValidatorService fileUploadsValidatorService)
         {
             _fileTypeService = fileTypeService;
+            _fileUploadsValidatorService = fileUploadsValidatorService;
         }
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataArchiveValidationService.cs:36]: Done.

I've not been converting constructors to use a primary constructor because I think it pollutes the PR quite a bit. I'd prefer to make the life of the reviewer easier over switching the constructor.

Maybe we can have someone just switch to primary constructors across the codebase in it's own ticket and get it done with - can't imagine it would take long and make PRs much easier to review than when such changes are intermingled. I'll create a ticket.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataArchiveValidationService.cs [hunk 13]
```diff
@@ -18,99 +23,274 @@ namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
...
+            // Check for duplicate data set filenames - because the bulk zip itself main contain duplicates!
+            dataSetNamesCsvEntries
+                .GroupBy(entry  => entry.BaseFilename)
+                .Where(group => group.Count() > 1)
+                .Select(group => group.Key)
+                .ForEach(duplicateFilename =>
+                {
+                    errors.Add(ValidationMessages
+                        .GenerateErrorDatasetNamesCsvFilenamesShouldBeUnique(duplicateFilename));
+                });
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataArchiveValidationService.cs:198]: I've made the change.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 14]
```diff
@@ -89,170 +143,177 @@ public class FileUploadsValidatorService : IFileUploadsValidatorService
...
+            string filename)
         {
             return _context
                 .ReleaseFiles
                 .Include(rf => rf.File)
                 .Where(rf => rf.ReleaseVersionId == releaseVersionId)
                 .ToList()
-                .Any(rf => String.Equals(rf.File.Filename, name, CurrentCultureIgnoreCase)
+                .Any(rf => string.Equals(rf.File.Filename, filename, CurrentCultureIgnoreCase)
                            && rf.File.Type == type);
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:207]: Moved the type check into the above `Where`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataArchiveValidationService.cs [hunk 15]
```diff
@@ -18,99 +23,274 @@ namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
...
+            var dataSetNamesCsvEntries = new List<(string BaseFilename, string Title)>();
+            while (!lastLine)
+            {
+                var filename = dataSetNamesCsvReader.GetField<string>("file_name");
+                var datasetName = dataSetNamesCsvReader.GetField<string>("dataset_name");
+
+                dataSetNamesCsvEntries.Add((BaseFilename: filename, Title: datasetName));
+
+                lastLine = !await dataSetNamesCsvReader.ReadAsync();
+            }
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataArchiveValidationService.cs:168]: I switched to using CsvUtils. This does mean that the code now requires to the headers to be in the CSV in a certain order, but I think that's fine. I've updated the associated `ValidationMessage` to make this clear.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Processor/Services/DataArchiveService.cs [hunk 16]
```diff
@@ -23,15 +23,14 @@ public async Task ExtractDataFiles(DataImport import)
...
+            await using var zipBlobFileStream = await _privateBlobStorageService
+                .StreamBlob(PrivateReleaseFiles, path);
             using var archive = new ZipArchive(zipBlobFileStream);
 
-            var file1 = archive.Entries[0];
-            var file2 = archive.Entries[1];
-            var dataFile = file1.Name.Contains(".meta.") ? file2 : file1;
-            var metadataFile = file1.Name.Contains(".meta.") ? file1 : file2;
+            var dataFile = archive.GetEntry(import.File.Filename);
+            var metaFile = archive.GetEntry(import.MetaFile.Filename);
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Processor/Services/DataArchiveService.cs:31]: I think you're right - good spot.

I've removed the `ToLower` calls from `ArchiveDataSetFile`. When the Processor extracts the files from the zip, they're given the FileId as their filename to be stored in blob storage, so the only thing we need to lowercase is the new File table entry's Filename, which we already do. So the `ArchiveDataSetFile` `ToLower` calls are unnecessary.

[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Processor/Services/DataArchiveService.cs:31]: I'll make a note on the ticket for the testers to look out for this.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs]: That's okay but I regularly format all my changed files and spend time wondering whether to pollute my own changes fixing prexisting formatting issues or undo them. It also doesn't make it easy to work out why/when a change has been made if it's attributed to another issue later on.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs]: I think that one's ok to leave. 🙂

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:138]: I've switched to using `[DisableRequestSizeLimit]`, and made the same change across the whole codebase.

I'll leave a note to test for large files on the ticket (although I'm sure they would anyway).

File: src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs]: I think you already know, but the original code had a method for detecting zips which included both types - I'm not sure why. And used MimeDetective for zips and left a comment saying it was a better detector for zips - no idea what that was based on. 

I didn't look into why - I just went along with the previous code.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs]: Good point.

We should get the code formatting sorted and then this problem of thing disappears.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Services/Interfaces/IFileTypeService.cs [hunk 17]
```diff
@@ -8,12 +8,12 @@ namespace GovUk.Education.ExploreEducationStatistics.Common.Services.Interfaces
 {
     public interface IFileTypeService
     {
-        Task<string> GetMimeType(IFormFile file);
```
[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Services/Interfaces/IFileTypeService.cs]: I didn't - but now you point it out I'd be inclined to keep the unit tests, even if it's really testing dependencies for the most part.
